src/envoy/http/service_control/client_cache_test.cc

```diff
@@ -268,6 +283,184 @@ TEST_F(ClientCacheQuotaResponseErrorTypeTest, ApiKeyInvalid) {
+
+    // Setup mock http call.
+    http_call_ = std::make_unique<MockHttpCall>();
+    check_call_factory_ = std::make_unique<MockHttpCallFactory>();
+    quota_call_factory_ = std::make_unique<MockHttpCallFactory>();
+    report_call_factory_ = std::make_unique<MockHttpCallFactory>();
+  }
+
+  void TearDown() override {
+    EXPECT_EQ(got_num_callbacks, want_num_callbacks);
+    ClientCacheTestBase::TearDown();
+  }
+
+  void injectFactoryMocks() {
+    cache_->check_call_factory_ = std::move(check_call_factory_);
+    cache_->quota_call_factory_ = std::move(quota_call_factory_);
+    cache_->report_call_factory_ = std::move(report_call_factory_);
+  }
+
+  NiceMock<Envoy::Tracing::MockSpan> mock_parent_span_;
+  std::unique_ptr<MockHttpCall> http_call_;
+
+  // Ownership of these is passed to ClientCache after calling
+  // `injectFactoryMocks`.
+  std::unique_ptr<MockHttpCallFactory> check_call_factory_;
+  std::unique_ptr<MockHttpCallFactory> quota_call_factory_;
+  std::unique_ptr<MockHttpCallFactory> report_call_factory_;
+
+  // Track the number of callbacks. This will ensure no async calls are left on
+  // destruction / tearDown.
+  int got_num_callbacks = 0;
+  int want_num_callbacks = 0;
+};
+
+class ClientCacheCheckHttpRequestTest : public ClientCacheHttpRequestTest {
+ public:
+  CheckRequest getValidCheckRequest() {
+    CheckRequest request;
+    request.set_service_name(kServiceName);
+    request.set_service_config_id(kServiceConfigId);
+    Operation* op = request.mutable_operation();
+    op->set_operation_id(kCheckOperationId);
+    op->set_operation_name("test_check_operation_name");
+    op->set_consumer_id("test-api-key");
+    return request;
+  }
+
+  CheckResponse getValidCheckResponse() {
+    CheckResponse response;
+    response.set_operation_id(kCheckOperationId);
+    response.set_service_config_id(kServiceConfigId);
+    return response;
+  }
+
+  void verifyCheckDone(const Code want_code, const Status& got_status,
+                       const CheckResponseInfo&) {
+    got_num_callbacks++;
+    EXPECT_EQ(got_status.code(), want_code)
+        << "Got message: " << got_status.message();
+  };
+};
+
+TEST_F(ClientCacheCheckHttpRequestTest, OneSuccessfulHttpCall) {
+  // Only 1 http call will be created.
+  HttpCall::DoneFunc http_done;
+  EXPECT_CALL(*check_call_factory_, createHttpCall(_, _, _))
+      .WillOnce(Invoke([this, &http_done](const Envoy::Protobuf::Message&,
+                                          Envoy::Tracing::Span&,
+                                          HttpCall::DoneFunc on_done) {
+        http_done = on_done;
+        return http_call_.get();
+      }));
+  EXPECT_CALL(*http_call_, call()).Times(1);
+  injectFactoryMocks();
+
+  // Run function under test.
+  const CheckRequest request = getValidCheckRequest();
+  cache_->callCheck(
+      request, mock_parent_span_,
+      absl::bind_front(&ClientCacheCheckHttpRequestTest::verifyCheckDone, this,
```

Comment:
can be use lambda function?  not easy to read with bind_front.

Reply:
I prefer this actually, it reduces code re-use. Lambda would add 4 duplicate lines of code to every test. This is pretty common google style I thought.

I could define a lambda in the test fixture class, but that seems weird.

Reply:
Hmm,  how about this lambda,  it is just one-line.

[this](const Status& got, const CheckResponseInfo&) { verifyCheckDone(Code::OK, got); }

Reply:
Done. More than 1 line because of clang format and also counting the number of callbacks.